src/Nethermind/Nethermind.Blockchain/Receipts/ReceiptsRecovery.cs

```diff
@@ -50,23 +55,98 @@ public ReceiptsRecoveryResult TryRecover(Block block, TxReceipt[] receipts, bool
-            receipt.Recipient = transaction.IsContractCreation ? null : transaction.To;
-
-            // how would it be in CREATE2?
-            receipt.ContractAddress = transaction.IsContractCreation && transaction.SenderAddress is not null ? ContractAddress.From(receipt.Sender, transaction.Nonce) : null;
-            receipt.GasUsed = receipt.GasUsedTotal - gasUsedBefore;
-            if (receipt.StatusCode != StatusCode.Success)
+            private readonly IReleaseSpec _releaseSpec;
+            private readonly Block _block;
+            private readonly bool _forceRecoverSender;
+            private readonly IEthereumEcdsa _ecdsa;
+
+            private long _gasUsedBefore = 0;
+            private int _transactionIndex = 0;
+
+            public RecoveryContext(IReleaseSpec releaseSpec, Block block, bool forceRecoverSender, IEthereumEcdsa ecdsa)
+            {
+                _releaseSpec = releaseSpec;
+                _block = block;
+                _forceRecoverSender = forceRecoverSender;
+                _ecdsa = ecdsa;
+            }
+
+            public void RecoverReceiptData(TxReceipt receipt)
+            {
+                if (_transactionIndex >= _block.Transactions.Length)
+                {
+                    throw new InvalidOperationException("Trying to recover more receipt that transaction");
+                }
+
+                Transaction transaction = _block.Transactions[_transactionIndex];
+
+                receipt.TxType = transaction.Type;
+                receipt.BlockHash = _block.Hash;
+                receipt.BlockNumber = _block.Number;
+                receipt.TxHash = transaction.Hash;
+                receipt.Index = _transactionIndex;
+                receipt.Sender = transaction.SenderAddress ?? (_forceRecoverSender ? _ecdsa.RecoverAddress(transaction, !_releaseSpec.ValidateChainId) : null);
+                receipt.Recipient = transaction.IsContractCreation ? null : transaction.To;
+
+                // how would it be in CREATE2?
+                receipt.ContractAddress = transaction.IsContractCreation && transaction.SenderAddress is not null ? ContractAddress.From(receipt.Sender, transaction.Nonce) : null;
+                receipt.GasUsed = receipt.GasUsedTotal - _gasUsedBefore;
+                if (receipt.StatusCode != StatusCode.Success)
+                {
+                    receipt.StatusCode = (receipt.Logs?.Length ?? 0) == 0 ? StatusCode.Failure : StatusCode.Success;
+                }
+
+                IncrementContext(receipt.GasUsedTotal);
+            }
+
+            public void RecoverReceiptData(ref TxReceiptStructRef receipt)
+            {
+                if (_transactionIndex >= _block.Transactions.Length)
+                {
+                    throw new InvalidOperationException("Trying to recover more receipt that transaction");
+                }
+
+                Transaction transaction = _block.Transactions[_transactionIndex];
```

Comment:
We would recover even if it is not needed.

Reply:
I'll add a skip.